src/main.rs

```diff
@@ -10,6 +10,46 @@ use libazureinit::{
 
 const VERSION: &str = env!("CARGO_PKG_VERSION");
 
+fn get_username(
+    imds_body: String,
+) -> Result<String, Box<dyn std::error::Error>> {
+    let username;
+    let password;
+
+    if match imds::get_provision_with_password(&imds_body) {
+        Ok(disable_passwd_auth) => disable_passwd_auth,
+        Err(_err) => {
+            return Err("Failed to get disable password authentication".into())
+        }
+    } {
+        // password authentication is disabled
+        let username_result = imds::get_username(imds_body.clone());
+        username = match username_result {
+            Ok(username) => username,
+            Err(_err) => return Err("Failed to get username".into()),
+        };
+    } else {
+        // password authentication is enabled
+        let ovf_body = media::read_ovf_env_to_string().unwrap();
+        let environment = media::parse_ovf_env(ovf_body.as_str()).unwrap();
+
+        username = environment
+            .provisioning_section
+            .linux_prov_conf_set
+            .username;
+        password = environment
+            .provisioning_section
+            .linux_prov_conf_set
+            .password;
+
+        if !password.is_empty() {
+            return Err("password is non-empty".into());
+        }
```

Comment:
Since you're only using `password` to check if it's empty you could replace this with
```suggestion
        if environment.provisioning_section.linux_prov_conf_set.password.is_empty() {
            return Err("password is non-empty".into());
        }
```

Then you could drop the `let password;` at the start of the function since it's not needed in the outer scope.

If you wanted, you could also drop the `let username;` as well and just return the username from both branches of the `if/else` rather than binding it to the variable.

Reply:
Done. Removed username/password variables, and reduced number of lines as much as possible.